File: CA_Server/Server/Game/Object/Item.cs
using Google.Protobuf.Protocol;
using System;
using System.Collections.Generic;
using System.Text;

namespace Server.Game
{
    public class Item : GameObject
    {
        public int ItemType { get; set; }

        public Item()
        {
            ObjectType = GameObjectType.Item;
        }

        public override void OnAttacked(GameObject attacker)
        {
            Despawn();
        }

        public void OnSpawn()
        {
            Console.WriteLine($"{Id} : Item Spawn ({CellPos.x}, {CellPos.y})");

            Room.Push(CheckCollision);
        }

        private void CheckCollision()
        {
            if (Room == null)
                return;

            // 플레이어가 존재하는지 확인
            List<GameObject> gameObjects = Room.FindAll(CellPos);
            if (gameObjects != null)
            {
                foreach (GameObject go in gameObjects)
                {
                    if (go.ObjectType == GameObjectType.Player)
                    {
                        Player p = go as Player;
                        CreatureState state = p.Info.PosInfo.State;
                        if (state == CreatureState.Idle || state == CreatureState.Moving)
                        {
                            p.GetItem(this);
                            Despawn();
                            return;
                        }

                    }
                }
            }

            // 0.1초마다 체크
            Room.PushAfter(100, CheckCollision);
        }

        private void Despawn()
        {
            Room.Push(Room.LeaveGame, Id);

            Console.WriteLine($"{Id} : Item Despawn");
        }
    }
}

File: CA_CppServer/GameServer/Item.cpp
#include "pch.h"
#include "Item.h"
#include "GameRoom.h"
#include "Player.h"

void Item::OnSpawn()
{
	GameRoomRef gameRoom = room.lock();
	if (gameRoom == nullptr)
		return;

	ItemRef mySelf = static_pointer_cast<Item>(shared_from_this());
	gameRoom->DoAsync([=]() { CheckCollision(mySelf); });
}

void Item::OnAttacked(GameObjectRef attacker)
{
	Despawn();
}

void Item::CheckCollision(ItemRef mySelf)
{
	if (_isExist == false)
		return;

	GameRoomRef gameRoom = room.lock();
	if (gameRoom == nullptr)
		return;

	// 플레이어가 존재하는지 확인
	vector<GameObjectRef> gameObjects = gameRoom->FindAllObjects(GetCellPos());
	for (GameObjectRef go : gameObjects)
	{
		if (go->objectType == Protocol::GameObjectType::PLAYER)
		{
			PlayerRef player = reinterpret_pointer_cast<Player>(go);
			Protocol::CreatureState state = player->info.posinfo().state();
			if (state == Protocol::CreatureState::IDLE || state == Protocol::CreatureState::MOVING)
			{
				player->GetItem(mySelf);
				Despawn();
				return;
			}
		}
	}

	// 0.1초마다 체크
	gameRoom->DoTimer(100, [=]() {	CheckCollision(mySelf); });
}

void Item::Despawn()
{
	GameRoomRef gameRoom = room.lock();
	if (gameRoom == nullptr)
		return;

	_isExist = false;
	gameRoom->DoAsync(&GameRoom::LeaveGame, info.objectid());
}
